File: wordNet/src/main/java/org/sqlunet/wordnet/loaders/SenseModule.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.sqlunet.wordnet.loaders;

import android.os.Parcelable;

import org.sqlunet.HasWordId;
import org.sqlunet.browser.TreeFragment;
import org.sqlunet.model.TreeFactory;
import org.sqlunet.treeview.model.TreeNode;
import org.sqlunet.wordnet.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Module for WordNet sense
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */

public class SenseModule extends SynsetModule
{
	/**
	 * Word id
	 */
	@Nullable
	private Long wordId;

	/**
	 * Constructor
	 *
	 * @param fragment fragment
	 */
	public SenseModule(@NonNull final TreeFragment fragment)
	{
		super(fragment);
	}

	@Override
	protected void unmarshal(final Parcelable pointer)
	{
		super.unmarshal(pointer);

		this.wordId = null;
		if (pointer instanceof HasWordId)
		{
			final HasWordId wordPointer = (HasWordId) pointer;
			this.wordId = wordPointer.getWordId();
		}
	}

	@Override
	public void process(@NonNull final TreeNode parent)
	{
		if (this.wordId != null && this.wordId != 0 && this.synsetId != null && this.synsetId != 0)
		{
			// anchor nodes
			final TreeNode synsetNode = TreeFactory.makeTextNode("Sense", false).addTo(parent);
			final TreeNode membersNode = TreeFactory.makeIconTextNode("Members", R.drawable.members, false).addTo(parent);

			// synset
			synset(this.synsetId, synsetNode, false);

			// members2
			members(this.synsetId, membersNode);

			// morph
			final TreeNode morphsNode = TreeFactory.makeTreeNode("Morphs", R.drawable.morph, false).addTo(parent);
			morphs(this.wordId, morphsNode);

			// relations and samples
			if (this.expand)
			{
				TreeFactory.makeHotQueryNode("Relations", R.drawable.ic_relations, false, new RelationsQuery(this.synsetId, this.wordId)).addTo(parent);
			}
			else
			{
				TreeFactory.makeQueryNode("Relations", R.drawable.ic_relations, false, new RelationsQuery(this.synsetId, this.wordId)).addTo(parent);
			}

			if (this.expand)
			{
				TreeFactory.makeHotQueryNode("Samples", R.drawable.sample, false, new SamplesQuery(this.synsetId)).addTo(parent);
			}
			else
			{
				TreeFactory.makeQueryNode("Samples", R.drawable.sample, false, new SamplesQuery(this.synsetId)).addTo(parent);
			}

			// special
			if (this.pos != null)
			{
				switch (this.pos)
				{
					case 'v':
						final TreeNode vframesNode = TreeFactory.makeTreeNode("Verb frames", R.drawable.verbframe, false).addTo(parent);
						final TreeNode vtemplatesNode = TreeFactory.makeTreeNode("Verb templates", R.drawable.verbtemplate, false).addTo(parent);
						vFrames(this.synsetId, this.wordId, vframesNode);
						vTemplates(this.synsetId, this.wordId, vtemplatesNode);
						break;

					case 'a':
						final TreeNode adjpositionsNode = TreeFactory.makeTreeNode("Adj positions", R.drawable.adjposition, false).addTo(parent);
						adjPosition(this.synsetId, this.wordId, adjpositionsNode);
						break;
				}
			}
			else
			{
				final TreeNode vframesNode = TreeFactory.makeTreeNode("Verb frames", R.drawable.verbframe, false).addTo(parent);
				final TreeNode vtemplatesNode = TreeFactory.makeTreeNode("Verb templates", R.drawable.verbtemplate, false).addTo(parent);
				final TreeNode adjpositionsNode = TreeFactory.makeTreeNode("Adj positions", R.drawable.adjposition, false).addTo(parent);
				vFrames(this.synsetId, this.wordId, vframesNode);
				vTemplates(this.synsetId, this.wordId, vtemplatesNode);
				adjPosition(this.synsetId, this.wordId, adjpositionsNode);
			}
		}
		else
		{
			TreeFactory.setNoResult(parent);
		}
	}
}

File: wordNet/src/main/java/org/sqlunet/wordnet/loaders/SenseModule.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.wordnet.loaders

import android.os.Parcelable
import org.sqlunet.HasWordId
import org.sqlunet.browser.TreeFragment
import org.sqlunet.model.TreeFactory.makeHotQueryLinkNode
import org.sqlunet.model.TreeFactory.makeHotQueryNode
import org.sqlunet.model.TreeFactory.makeHotQueryTreeNode
import org.sqlunet.model.TreeFactory.makeIconTextNode
import org.sqlunet.model.TreeFactory.makeLinkHotQueryTreeNode
import org.sqlunet.model.TreeFactory.makeQueryTreeNode
import org.sqlunet.model.TreeFactory.makeTextNode
import org.sqlunet.model.TreeFactory.makeTreeNode
import org.sqlunet.model.TreeFactory.setNoResult
import org.sqlunet.treeview.control.Link
import org.sqlunet.treeview.model.TreeNode
import org.sqlunet.wordnet.R

/**
 * Module for WordNet sense
 *
 * @param fragment fragment
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
class SenseModule(fragment: TreeFragment) : SynsetModule(fragment) {

    /**
     * Word id
     */
    private var wordId: Long? = null

    override fun unmarshal(pointer: Parcelable) {
        super.unmarshal(pointer)
        wordId = null
        if (pointer is HasWordId) {
            val wordPointer = pointer as HasWordId
            wordId = wordPointer.getWordId()
        }
    }

    override fun process(node: TreeNode) {
        if (wordId != null && wordId != 0L && synsetId != null && synsetId != 0L) {
            // anchor nodes
            val synsetNode = makeTextNode(senseLabel, false).addTo(node)
            val membersNode = makeIconTextNode(membersLabel, R.drawable.members, false).addTo(node)

            // synset
            synset(synsetId!!, synsetNode, false)

            // member set
            members(synsetId!!, membersNode)

            // morph
            val morphsNode = makeTreeNode(morphsLabel, R.drawable.morph, false).addTo(node)
            morphs(wordId!!, morphsNode)

            // relations
            if (expand) {
                val link: Link = RelationLink(synsetId!!, maxRecursion, fragment)
                makeLinkHotQueryTreeNode(relationsLabel, R.drawable.ic_relations, false, RelationsQuery(synsetId!!, wordId!!), link, R.drawable.ic_link_relation).addTo(node)
            } else {
                makeQueryTreeNode(relationsLabel, R.drawable.ic_relations, false, RelationsQuery(synsetId!!, wordId!!)).addTo(node)
            }

            // samples
            if (expand) {
                makeHotQueryTreeNode(samplesLabel, R.drawable.sample, false, SamplesQuery(synsetId!!)).addTo(node)
            } else {
                makeQueryTreeNode(samplesLabel, R.drawable.sample, false, SamplesQuery(synsetId!!)).addTo(node)
            }

            // usages
            makeHotQueryNode(usagesLabel, R.drawable.usage, false, UsagesQuery(synsetId!!)).addTo(node)

            // externals
            makeHotQueryLinkNode(iliLabel, R.drawable.ili, false, IliQuery(synsetId!!)).addTo(node)
            makeHotQueryLinkNode(wikidataLabel, R.drawable.wikidata, false, WikidataQuery(synsetId!!)).addTo(node)

            // special
            if (pos != null) {
                when (pos!!) {
                    'v' -> {
                        val vframesNode = makeTreeNode(verbFramesLabel, R.drawable.verbframe, false).addTo(node)
                        val vtemplatesNode = makeTreeNode(verbTemplatesLabel, R.drawable.verbtemplate, false).addTo(node)
                        vFrames(synsetId!!, wordId!!, vframesNode)
                        vTemplates(synsetId!!, wordId!!, vtemplatesNode)
                    }

                    'a' -> {
                        val adjpositionsNode = makeTreeNode(adjPositionsLabel, R.drawable.adjposition, false).addTo(node)
                        adjPosition(synsetId!!, wordId!!, adjpositionsNode)
                    }
                }
            } else {
                val vframesNode = makeTreeNode(verbFramesLabel, R.drawable.verbframe, false).addTo(node)
                val vtemplatesNode = makeTreeNode(verbTemplatesLabel, R.drawable.verbtemplate, false).addTo(node)
                val adjpositionsNode = makeTreeNode(adjPositionsLabel, R.drawable.adjposition, false).addTo(node)
                vFrames(synsetId!!, wordId!!, vframesNode)
                vTemplates(synsetId!!, wordId!!, vtemplatesNode)
                adjPosition(synsetId!!, wordId!!, adjpositionsNode)
            }
        } else {
            setNoResult(node)
        }
    }
}
